modulemd/v2/modulemd-defaults.c

```diff
@@ -0,0 +1,309 @@
+                           GError **error)
+{
+  g_assert_true (MODULEMD_IS_DEFAULTS (self));
+
+  if (!mdversion)
+    mdversion = MD_DEFAULTS_VERSION_LATEST;
+
+  if (mdversion > MD_DEFAULTS_VERSION_LATEST)
+    {
+      g_set_error (error,
+                   MODULEMD_ERROR,
+                   MODULEMD_ERROR_UPGRADE,
+                   "Unknown metadata version for upgrade: %" PRIu64 ".",
+                   mdversion);
+      return NULL;
+    }
+
+  if (modulemd_defaults_get_mdversion (self) == mdversion)
+    {
+      /* Already at this version, just copy it and return that */
+      return modulemd_defaults_copy (self);
+    }
+
+  return NULL;
+}
+
+
+guint64
+modulemd_defaults_get_mdversion (ModulemdDefaults *self)
+{
+  ModulemdDefaultsClass *klass;
+
+  g_return_val_if_fail (self && MODULEMD_IS_DEFAULTS (self), 0);
+
+  klass = MODULEMD_DEFAULTS_GET_CLASS (self);
+  g_return_val_if_fail (klass->get_mdversion, 0);
+
+  return klass->get_mdversion (self);
+}
+
+
+static void
+modulemd_defaults_set_module_name (ModulemdDefaults *self,
+                                   const gchar *module_name)
+{
+  g_return_if_fail (MODULEMD_IS_DEFAULTS (self));
+
+  /* It is a coding error if we ever get a NULL name here */
+  g_return_if_fail (module_name);
+
+  /* It is a coding error if we ever get the default name here */
+  g_return_if_fail (g_strcmp0 (module_name, DEF_DEFAULT_NAME_STRING));
+
+  ModulemdDefaultsPrivate *priv =
+    modulemd_defaults_get_instance_private (self);
+
+  g_clear_pointer (&priv->module_name, g_free);
+  priv->module_name = g_strdup (module_name);
+
+  g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_MODULE_NAME]);
+}
+
+
+const gchar *
+modulemd_defaults_get_module_name (ModulemdDefaults *self)
+{
+  ModulemdDefaultsPrivate *priv =
+    modulemd_defaults_get_instance_private (self);
+
+  return priv->module_name;
+}
+
+
+static void
+modulemd_defaults_get_property (GObject *object,
+                                guint prop_id,
+                                GValue *value,
+                                GParamSpec *pspec)
+{
+  ModulemdDefaults *self = MODULEMD_DEFAULTS (object);
+
+  switch (prop_id)
+    {
+    case PROP_MODULE_NAME:
+      g_value_set_string (value, modulemd_defaults_get_module_name (self));
+      break;
+
+    case PROP_MDVERSION:
+      g_value_set_uint64 (value, modulemd_defaults_get_mdversion (self));
+      break;
+
+    default: G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+modulemd_defaults_set_property (GObject *object,
+                                guint prop_id,
+                                const GValue *value,
+                                GParamSpec *pspec)
+{
+  ModulemdDefaults *self = MODULEMD_DEFAULTS (object);
+
+  switch (prop_id)
+    {
+    case PROP_MODULE_NAME:
+      modulemd_defaults_set_module_name (self, g_value_get_string (value));
+      break;
+
+    case PROP_MDVERSION:
```

Comment:
Maybe instead return an explicit error that while `MDVERSION` is valid, it can't be set?

Reply:
That's exactly what this does. It falls through to the default clause which does `G_OBJECT_WARN_INVALID_PROPERTY_ID` which means "this property isn't valid for this call".

Reply:
Right, I meant to suggest not using the fallthrough and returning an explicit "This attribute is not supposed to be modified" error message.
But if you think the standard error suffices, fine with me.